Add parsing logic for `Tuple[()]` annotation

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58340 Add parsing logic for `Tuple[()]` annotation**

Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

Commit:
Add parsing logic for `Tuple[()]` annotation

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28459502).

jamesr66a commented on torch/_jit_internal.py:
```diff
@@ -783,7 +788,7 @@ def _get_overloaded_methods(method, mod_class):
 
 def is_tuple(ann) -> bool:
-    if ann is Tuple:
+    if ann is Tuple or ann is tuple:
```
There seem to be some action-at-a-distance test failures happening. guessing but maybe this would work?

```
if ann is Tuple or getattr(ann, '__origin__', None) is tuple:
```

(Note i'm not 100% sure why the change here is needed. where are these type annotations that are not the `typing` classes coming from?

ansley replied:
Turns out that "fixing" code without reading the line directly below it is Not Good™. Uh, I've got half of "move fast, break things" down?

jamesr66a reviewed (commented):
Looks good but test failures indicate something wonky going on

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28459502).

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]

jamesr66a reviewed (approved):
Great, make sure tests pass and fix the small lint nit

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28459502).

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28459502).

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]

Commit:
Update on "Add parsing logic for `Tuple[()]` annotation"


Differential Revision: [D28459502](https://our.internmc.facebook.com/intern/diff/D28459502)

[ghstack-poisoned]
